DirectProgramming/C++SYCL_FPGA/ReferenceDesigns/board_test/src/usm_speed.hpp

```diff
@@ -0,0 +1,177 @@
+class USMMemCopy;
+class USMMemRead;
+class USMMemWrite;
+
+using long8 = sycl::vec<long, 8>;
+
+enum USMTest { MEMCOPY, READ, WRITE };
+constexpr size_t kIterations = 1;
+constexpr size_t kNumBytes = 1024 * 1024 * 1024;
+constexpr size_t kNumItems = kNumBytes / sizeof(long8);
+
+// MEMCOPY test: launches a kernel to copy data from one USM pointer to another.
+sycl::event memcopy_kernel(sycl::queue &q, long8 *in, long8 *out) {
+  return q.single_task<USMMemCopy>([=]() [[intel::kernel_args_restrict]] {
+    sycl::host_ptr<long8> in_h(in);
+    sycl::host_ptr<long8> out_h(out);
+    for (size_t i = 0; i < kNumItems; i++) {
+      out_h[i] = in_h[i];
+    }
+  });
+}
+
+// READ test: launches a kernel to read data from a USM pointer, sum it up, and
+// store to an output pointer.
+sycl::event read_kernel(sycl::queue &q, long8 *in, long8 *out) {
+  return q.single_task<USMMemRead>([=]() {
+    sycl::host_ptr<long8> in_h(in);
+    sycl::host_ptr<long8> out_h(out);
+    long8 sum{0};
+    for (size_t i = 0; i < kNumItems; i++) {
+      sum += in_h[i];
+    }
+    // This prevents the reads from being optimized away
+    out_h[0] = sum;
+  });
+}
+
+// WRITE test: launches a kernel to write data to a USM pointer.
+sycl::event write_kernel(sycl::queue &q, long8 *in, long8 *out) {
+  return q.single_task<USMMemWrite>([=]() {
+    sycl::host_ptr<long8> out_h(out);
+    long8 answer{5};
+    for (size_t i = 0; i < kNumItems; i++) {
+      out_h[i] = answer;
+    }
+  });
+}
+
+// Function to check the answer of the memcopy, read or write test.
+bool verify(long8 *in, long8 *out, USMTest test) {
+
+  // The write kernel writes a known value to every index of "out".
+  long8 write_answer{5};
+
+  // The read kernel calculates a sum of all the values at "in" and stores it
+  // at out[0]. First calculate a reference to compare to. The rest of the
+  // indices should retain their intial value, 0. So, also define a zero vector
+  // to compare to.
+  long8 zero_vec{0};
+  long8 read_answer{0};
+  for (int i = 0; i < kNumItems; i++) {
+    read_answer += in[i];
```

Comment:
Logic like summing the elements in the array is only relevant for the READ case, so I think this entire for loop should be under the READ case in the below switch statement, since the code doesn't need to execute in any other case. 

In fact, I'd say that none of these, zero_vec, read_answer, write_answer should exist outside of the switch statement below that filters by the "test" variable. I think you can just assign to reference directly!

Reply:
I tried implementing this, but noticed that moving the initialization of read_answer to within the switch statement (and thus within the loop on line 65) means it will be executed for every element we are checking (i.e., from ~2*kNumItems iterations to ~kNumItems^2 iterations). Since kNumItems is quite large, this caused a significant increase in emulation runtime. 

For now, I've moved the write_answer and zero values to exist only within their respective case in the switch statement; I'll see if I can find a better solution for the read variable.

Reply:
Ah, I see! Glad you found the significant increase in runtime, yeah putting the calculation in the for loop is not great actually.

One thing that may work is having a reference array instead of a reference element below.

something like 

```
sycl::ulong8 expected[kNumItems]

switch(test){
  case MEMCOPY:
    // copy "in" into expected
  case READ:
   // set expected to zero, and set the first element to be the sum
  case WRITE:
  // set expected to fixed value
}
```

Then, in the main for loop, you can remove the switch statement, and simply compare out and the expected array you created earlier directly.

Reply:
I've implemented this in the sample -- and was able to do some further cleanup based on this. The verifier doesn't necessarily need to know which test is running; it just takes two pointers to compare (one is the USM pointer, the other is the reference array). Then, the run_test function can just populate the reference array in the switch statement where it selects the test.